File: packages/intelligence-delivery-python-core/src/_delivery/registry.py
"""Private asyncio registry with immutable snapshots and one shared refresh."""

import asyncio
import logging
from dataclasses import dataclass
from datetime import UTC, datetime
from time import monotonic, time
from typing import Literal

from copilotkit_intelligence import Intelligence, LearnedSkillsError, LearnedSkillsErrorCode

from .config import resolve_config
from .snapshot import VerifiedSnapshot, invalid_snapshot, validate_snapshot

logger = logging.getLogger(__name__)
_TRANSIENT = frozenset({"NETWORK_ERROR", "TIMEOUT", "INVALID_SNAPSHOT", "UNSUPPORTED_SERVER"})


@dataclass(frozen=True, slots=True)
class ErrorStatus:
    code: LearnedSkillsErrorCode
    message: str
    retryable: bool


@dataclass(frozen=True, slots=True)
class Status:
    initialized: bool
    revision: str | None
    mode: Literal["latest", "pinned"]
    last_checked_at: str | None
    stale: bool
    last_error: ErrorStatus | None


def _consume(task: asyncio.Task[VerifiedSnapshot]) -> None:
    """Observe abandoned task failures without logging exception content."""
    if not task.cancelled():
        task.exception()


class Registry:
    """Internal registry, vendored into each framework's private delivery namespace."""

    def __init__(
        self,
        *,
        client: Intelligence | None = None,
        api_key: str | None = None,
        api_url: str | None = None,
        container_id: str | None = None,
        revision: str | None = None,
        freshness_window: float = 5,
        request_timeout: float = 5,
        debug: bool = False,
    ) -> None:
        self._config = resolve_config(
            client=client,
            api_key=api_key,
            api_url=api_url,
            container_id=container_id,
            revision=revision,
            freshness_window=freshness_window,
            request_timeout=request_timeout,
            debug=debug,
        )
        self._snapshot: VerifiedSnapshot | None = None
        self._inflight: asyncio.Task[VerifiedSnapshot] | None = None
        self._checked: float | None = None
        self._checked_at: str | None = None
        self._stale = False
        self._error: ErrorStatus | None = None
        self._blocked: ErrorStatus | None = None
        self._closed = False

    @property
    def status(self) -> Status:
        return Status(
            self._snapshot is not None,
            self._snapshot.revision if self._snapshot else None,
            "pinned" if self._config.revision is not None else "latest",
            self._checked_at,
            self._stale,
            self._error,
        )

    async def initialize(self) -> None:
        await self.acquire_snapshot()

    async def acquire_snapshot(self) -> VerifiedSnapshot:
        if self._closed:
            raise LearnedSkillsError("INVALID_CONFIG", False)
        if self._inflight is None:
            if (
                self._snapshot is not None
                and self._blocked is None
                and self._checked is not None
                and monotonic() - self._checked < self._config.freshness_window
            ):
                return self._snapshot
            self._inflight = asyncio.create_task(self._refresh())
            self._inflight.add_done_callback(self._finished)
        # One cancelled invocation must not cancel another invocation's refresh.
        return await asyncio.shield(self._inflight)

    def _finished(self, task: asyncio.Task[VerifiedSnapshot]) -> None:
        if self._inflight is task:
            self._inflight = None
        _consume(task)

    async def aclose(self) -> None:
        self._closed = True
        if self._inflight is not None:
            self._inflight.cancel()
            await asyncio.gather(self._inflight, return_exceptions=True)
        if self._config.owns_client:
            await self._config.client.aclose()

    async def _load(self, started: float) -> VerifiedSnapshot:
        response = await self._config.client.get_learned_skills_snapshot(
            container_id=self._config.container_id,
            revision=self._config.revision,
            if_none_match=self._snapshot.etag if self._snapshot else None,
            request_timeout=self._config.request_timeout,
        )
        if not isinstance(response, dict):
            raise invalid_snapshot()
        if response.get("status") == "unchanged":
            if (
                self._snapshot is None
                or response.get("revision") != self._snapshot.revision
                or response.get("etag") != self._snapshot.etag
            ):
                raise invalid_snapshot()
            return self._snapshot
        if self._config.revision is not None and response.get("revision") != self._config.revision:
            raise invalid_snapshot()
        # Capture metadata and owned bytes before handing work to another thread.
        # A caller-owned response dictionary must not change an exact pin in flight.
        raw = response.get("bytes")
        if not isinstance(raw, (bytes, bytearray)):
            raise invalid_snapshot()
        captured = {
            "status": response.get("status"),
            "bytes": bytes(raw),
            "revision": response.get("revision"),
            "etag": response.get("etag"),
            "contentType": response.get("contentType"),
        }
        remaining = self._config.request_timeout - (monotonic() - started)
        if remaining <= 0:
            raise LearnedSkillsError("TIMEOUT", True)
        async with asyncio.timeout(remaining):
            snapshot = await asyncio.to_thread(validate_snapshot, captured)
        if self._config.revision is not None and snapshot.revision != self._config.revision:
            raise invalid_snapshot()
        return snapshot

    async def _refresh(self) -> VerifiedSnapshot:
        started = monotonic()
        try:
            # The canonical client owns the HTTP deadline. A second timer would
            # race confirmed denial against transport cleanup. Validation uses
            # only the remaining portion of this same invocation budget.
            snapshot = await self._load(started)
            self._snapshot = snapshot
            self._checked = monotonic()
            self._checked_at = (
                datetime.fromtimestamp(time(), UTC)
                .isoformat(timespec="milliseconds")
                .replace("+00:00", "Z")
            )
            self._stale = False
            self._error = self._blocked = None
            self._debug("checked", started)
            return snapshot
        except asyncio.CancelledError:
            raise
        except Exception as cause:
            error = (
                cause
                if isinstance(cause, LearnedSkillsError)
                else LearnedSkillsError(
                    "TIMEOUT" if isinstance(cause, TimeoutError) else "NETWORK_ERROR", True, cause
                )
            )
            info = ErrorStatus(error.code, error.message, error.retryable)
            if error.code not in _TRANSIENT:
                self._blocked = info
            self._error = self._blocked if self._blocked is not None else info
            self._debug("failed", started)
            if self._blocked is not None:
                self._stale = False
                blocked = self._blocked
                raise LearnedSkillsError(blocked.code, blocked.retryable, error.cause) from None
            if self._snapshot is not None:
                self._stale = True
                return self._snapshot
            raise error from None

    def _debug(self, event: str, started: float) -> None:
        if self._config.debug:
            logger.debug(
                "Learned skills %s container=%s revision=%s duration_ms=%.1f error_code=%s",
                event,
                self._config.container_id,
                self._snapshot.revision if self._snapshot else None,
                (monotonic() - started) * 1000,
                self._error.code if self._error else None,
            )

File: packages/runtime/src/v2/runtime/intelligence-platform/skill-registry/registry.ts
import { createHash } from "node:crypto";
import { resolveRegistryConfig } from "./config";
import type { SingleContainerConfig, SkillRegistryOptions } from "./config";
import { SkillDeliveryError, invalidSnapshot } from "./errors";
import type { SkillDeliveryErrorCode } from "./errors";
import { validateSnapshot } from "./snapshot";
import type { VerifiedSnapshot } from "./snapshot";

export interface SkillRegistryStatus {
  /** Present only for the explicit containers interface. */
  readonly containers?: readonly (SkillRegistryStatus & {
    readonly id: string;
  })[];
  readonly initialized: boolean;
  readonly revision?: string;
  readonly mode: "latest" | "pinned";
  /** ISO 8601 timestamp of the last successful 200 or matching 304 check. */
  readonly lastCheckedAt?: string;
  readonly stale: boolean;
  readonly lastError?: Readonly<{
    code: SkillDeliveryErrorCode;
    message: string;
    retryable: boolean;
  }>;
}

function transient(error: SkillDeliveryError): boolean {
  return [
    "NETWORK_ERROR",
    "TIMEOUT",
    "INVALID_SNAPSHOT",
    "UNSUPPORTED_SERVER",
  ].includes(error.code);
}
function deliveryError(error: unknown): SkillDeliveryError {
  if (error instanceof SkillDeliveryError) return error;
  return new SkillDeliveryError(
    error instanceof Error && error.name === "TimeoutError"
      ? "TIMEOUT"
      : "NETWORK_ERROR",
    true,
    error,
  );
}

/** Internal registry shared by framework adapters; public API is framework-native. */
class SingleContainerRegistry {
  readonly #config: SingleContainerConfig;
  #snapshot?: VerifiedSnapshot;
  #inFlight?: Promise<VerifiedSnapshot>;
  #lastCheckedAt?: number;
  #stale = false;
  #lastError?: SkillDeliveryError;
  #blocked?: SkillDeliveryError;

  constructor(config: SingleContainerConfig) {
    this.#config = config;
  }

  async initialize(): Promise<void> {
    await this.acquireSnapshot();
  }

  /** Capture exactly one immutable snapshot before model work begins. */
  acquireSnapshot(): Promise<VerifiedSnapshot> {
    if (this.#inFlight) return this.#inFlight;
    if (
      this.#snapshot &&
      !this.#blocked &&
      this.#lastCheckedAt !== undefined &&
      Date.now() - this.#lastCheckedAt < this.#config.freshnessWindowMs
    ) {
      return Promise.resolve(this.#snapshot);
    }
    // Defer work one microtask so reentrant and concurrent callers observe the
    // same promise before an injected client can start its request.
    this.#inFlight = Promise.resolve()
      .then(() => this.#refresh())
      .finally(() => {
        this.#inFlight = undefined;
      });
    return this.#inFlight;
  }

  get status(): SkillRegistryStatus {
    const error = this.#lastError;
    return Object.freeze({
      initialized: this.#snapshot !== undefined,
      ...(this.#snapshot ? { revision: this.#snapshot.revision } : {}),
      mode: this.#config.revision === undefined ? "latest" : "pinned",
      ...(this.#lastCheckedAt !== undefined
        ? { lastCheckedAt: new Date(this.#lastCheckedAt).toISOString() }
        : {}),
      stale: this.#stale,
      ...(error
        ? {
            lastError: Object.freeze({
              code: error.code,
              message: error.message,
              retryable: error.retryable,
            }),
          }
        : {}),
    });
  }

  async #refresh(): Promise<VerifiedSnapshot> {
    const startedAt = Date.now();
    const controller = new AbortController();
    let timer: ReturnType<typeof setTimeout> | undefined;
    const deadline = new Promise<never>((_, reject) => {
      timer = setTimeout(() => {
        const error = new SkillDeliveryError("TIMEOUT", true);
        controller.abort(error);
        // Give a canonical client that already received denial headers one
        // event-loop turn to report that denial before the watchdog wins.
        timer = setTimeout(() => reject(error), 0);
      }, this.#config.requestTimeoutMs);
    });
    const replacement = async (): Promise<VerifiedSnapshot> => {
      const response = await this.#config.client.getLearnedSkillsSnapshot({
        containerId: this.#config.containerId,
        ...(this.#config.revision !== undefined
          ? { revision: this.#config.revision }
          : {}),
        ...(this.#snapshot ? { ifNoneMatch: this.#snapshot.etag } : {}),
        signal: controller.signal,
      });
      controller.signal.throwIfAborted();
      if (!response || typeof response !== "object") throw invalidSnapshot();
      if (response.status === "unchanged") {
        if (
          !this.#snapshot ||
          response.etag !== this.#snapshot.etag ||
          response.revision !== this.#snapshot.revision
        )
          throw invalidSnapshot();
        return this.#snapshot;
      }
      if (
        this.#config.revision !== undefined &&
        response.revision !== this.#config.revision
      )
        throw invalidSnapshot();
      return validateSnapshot(response, controller.signal);
    };
    try {
      const snapshot = await Promise.race([replacement(), deadline]);
      // An abandoned parse never writes registry state: only the winning
      // complete replacement reaches this assignment.
      this.#snapshot = snapshot;
      this.#lastCheckedAt = Date.now();
      this.#stale = false;
      this.#lastError = undefined;
      this.#blocked = undefined;
      this.#debug("refresh-success", startedAt);
      return snapshot;
    } catch (cause) {
      const error = deliveryError(cause);
      if (!transient(error)) {
        this.#blocked = error;
        this.#stale = false;
        this.#lastError = error;
        this.#debug("refresh-denied", startedAt, error.code);
        throw error;
      }
      // A later unreachable server cannot undo an already confirmed denial.
      if (this.#blocked) {
        this.#lastError = this.#blocked;
        this.#debug("refresh-failed", startedAt, error.code);
        throw this.#blocked;
      }
      this.#lastError = error;
      this.#stale = this.#snapshot !== undefined;
      this.#debug("refresh-failed", startedAt, error.code);
      if (this.#snapshot) return this.#snapshot;
      throw error;
    } finally {
      if (timer !== undefined) clearTimeout(timer);
    }
  }

  #debug(
    event: string,
    startedAt: number,
    errorCode?: SkillDeliveryErrorCode,
  ): void {
    if (!this.#config.debug) return;
    console.debug("CopilotKit learned skills", {
      event,
      containerId: this.#config.containerId,
      revision: this.#snapshot?.revision,
      durationMs: Date.now() - startedAt,
      ...(errorCode ? { errorCode } : {}),
    });
  }
}

/** Compose independent container caches into one immutable invocation snapshot. */
export class SkillRegistry {
  readonly #sources: readonly {
    readonly id: string;
    readonly registry: SingleContainerRegistry;
  }[];
  readonly #multiple: boolean;
  #snapshot?: VerifiedSnapshot;
  #parts?: readonly VerifiedSnapshot[];
  #inFlight?: Promise<VerifiedSnapshot>;

  constructor(options: SkillRegistryOptions = {}) {
    const config = resolveRegistryConfig(options);
    this.#multiple = config.containers !== undefined;
    this.#sources =
      config.containers !== undefined
        ? config.containers.map(({ id, revision }) => ({
            id,
            registry: new SingleContainerRegistry(
              Object.freeze({
                client: config.client,
                freshnessWindowMs: config.freshnessWindowMs,
                requestTimeoutMs: config.requestTimeoutMs,
                debug: config.debug,
                containerId: id,
                ...(revision !== undefined ? { revision } : {}),
              }),
            ),
          }))
        : [
            {
              id: config.containerId!,
              registry: new SingleContainerRegistry({
                client: config.client,
                containerId: config.containerId!,
                revision: config.revision,
                freshnessWindowMs: config.freshnessWindowMs,
                requestTimeoutMs: config.requestTimeoutMs,
                debug: config.debug,
              }),
            },
          ];
  }

  /** Load every configured container before agent initialization completes. */
  async initialize(): Promise<void> {
    await this.acquireSnapshot();
  }

  /** Capture every container once; a failed source cannot yield a partial catalog. */
  acquireSnapshot(): Promise<VerifiedSnapshot> {
    if (!this.#multiple) return this.#sources[0].registry.acquireSnapshot();
    if (this.#inFlight) return this.#inFlight;
    this.#inFlight = Promise.all(
      this.#sources.map(({ registry }) => registry.acquireSnapshot()),
    )
      .then((parts) => {
        if (
          this.#snapshot &&
          parts.every((part, index) => part === this.#parts?.[index])
        ) {
          return this.#snapshot;
        }
        const identity = createHash("sha256")
          .update(
            JSON.stringify(
              parts.map((part, index) => [
                this.#sources[index].id,
                part.revision,
                part.etag,
              ]),
            ),
          )
          .digest("hex");
        const skills = Object.freeze(
          parts.flatMap((part, index) =>
            part.skills.map((skill) =>
              Object.freeze({
                ...skill,
                name: `${encodeURIComponent(this.#sources[index].id)}/${skill.name}`,
              }),
            ),
          ),
        );
        this.#parts = parts;
        this.#snapshot = Object.freeze({
          revision: `multi:${identity}`,
          etag: `"${identity}"`,
          skills,
        });
        return this.#snapshot;
      })
      .finally(() => {
        this.#inFlight = undefined;
      });
    return this.#inFlight;
  }

  /** Report each container's real revision; no aggregate server revision exists. */
  get status(): SkillRegistryStatus {
    if (!this.#multiple) return this.#sources[0].registry.status;
    const containers = Object.freeze(
      this.#sources.map(({ id, registry }) =>
        Object.freeze({ id, ...registry.status }),
      ),
    );
    const error =
      containers.find(
        (source) => source.lastError && !source.lastError.retryable,
      )?.lastError ?? containers.find((source) => source.lastError)?.lastError;
    const timestamps = containers.map((source) => source.lastCheckedAt);
    return Object.freeze({
      initialized: this.#snapshot !== undefined,
      mode: containers.every((source) => source.mode === "pinned")
        ? "pinned"
        : "latest",
      stale: containers.some((source) => source.stale),
      ...(timestamps.every((timestamp) => timestamp !== undefined)
        ? { lastCheckedAt: [...timestamps].sort()[0] }
        : {}),
      ...(error ? { lastError: error } : {}),
      containers,
    });
  }
}
